### PR TITLE
First pass at some augmentations

### PR DESCRIPTION
**Pull request by ksanjeevan**
Following up on #44:

### Have
- `class StretchSpecTime`: spectrogram time stretching
- `class GaussianNoise`: additive Gaussian noise to the spectrogram
- `class FrequencyMasking` and `class TimeMasking`: Implementing the masking from [SpecAugment](https://arxiv.org/pdf/1904.08779.pdf)
### Not yet / needs more discussion
- `PitchShift`...
- Make sure all the naming follows the latest convention, e.g. change to `StretchTime`

### To do
- Write tests
- Add more doc
- Additional transforms

### CODE REVIEW

**Comment by ksanjeevan** [comment]
On the masking: Classes have boolean `same_batch` in case user wants to apply random masks per element in the batch and channel, or all of them the same. Example where `same_batch=True` for `FrequencyMasking` but `same_batch=False` on `TimeMasking`:
grid

I was mentioning in #44 that this can be easily modified so the `max_value` parameter can be a tensor of sequence lengths. If a user is working with different sequence lengths, seems like it could be a good use case for wanting example independent masking?

File: torchaudio_contrib/augmentation.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        param_str = '(scale={})'.format(self.scale)
+        return self.__class__.__name__ + param_str
+
+
+def apply_masking(spect, max_value, mask_value, axis):
+
+    assert axis in [2,3]
+
+    v = torch.rand(spect.shape[:2])*max_value
+    v_0 = torch.rand(spect.shape[:2])*(spect.size(axis) - v)
```
Comment (keunwoochoi) [inline torchaudio_contrib/augmentation.py]: 75 and 76 -- pep8.

File: torchaudio_contrib/augmentation.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            return spect, noise
+
+    def __repr__(self):
+        param_str = '(scale={})'.format(self.scale)
+        return self.__class__.__name__ + param_str
+
+
+def apply_masking(spect, max_value, mask_value, axis):
+
+    assert axis in [2,3]
```
Comment (keunwoochoi) [inline torchaudio_contrib/augmentation.py]: i recently changed my habit, too, to prefer `raise` to `assert` for some reasons. reference: https://stackoverflow.com/questions/20076182/what-are-the-implications-of-using-assert-vs-raise-exception/20076295

File: torchaudio_contrib/augmentation.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    Args:
+        scale (float or tensor): standard deviation of the distribution.
+    """
+
+    def __init__(self, scale=1.):
+        super(GaussianNoise, self).__init__()
+        self.scale = torch.as_tensor(scale)
+        self.loc = torch.tensor(0.)
+
+    def forward(self, spect, scale=None):
```
Comment (keunwoochoi) [inline torchaudio_contrib/augmentation.py]: `spect` should be `mag_specgrams` or something like that?

File: torchaudio_contrib/augmentation.py (hunk 1 above)
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: and some other places, too!

File: torchaudio_contrib/augmentation.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+            noise = torch.normal(self.loc.expand(spect.shape), scale.expand(spect.shape)) 
+            return spect, noise
+
+    def __repr__(self):
+        param_str = '(scale={})'.format(self.scale)
+        return self.__class__.__name__ + param_str
+
+
+def apply_masking(spect, max_value, mask_value, axis):
```
Comment (keunwoochoi) [inline torchaudio_contrib/augmentation.py]: the name seems slight too general. probably something like `mask_random_region_along_axis`..but shorter? :D

Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: also i think we need little discussion at some point about if this masking would be always with zeros (which i think people would assume) or with a number as implemented here. it is more generalizable here, but because if the use-case is always with zeros, i'd expect after the masking the masked region is not valid values with zeros, but just doesn't exist at all (i.e., the output is smaller in size) so that computation could be more efficient.

could be tricky for us, maybe we also need to invent a slightly different name, because for `SpecAug`, actually, the current implementation can make more sense. overall, i'm not 100% sure about this but let's think about it.

File: torchaudio_contrib/augmentation.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        spect[:,:,:,mask_start:mask_end] = mask_value
+    else:
+        raise ValueError('Only Frequency and Time masking is supported')
+
+    return spect
+
+
+class _AxisMasking(_ModuleNoStateBuffers):
+
+    def __init__(self, max_value, axis, same_batch):
```
Comment (keunwoochoi) [inline torchaudio_contrib/augmentation.py]: `same_batch` --> something more intuitive? I couldn't get what it means from the name.

File: torchaudio_contrib/augmentation.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    def __init__(self, rate=1., hop_len=512, num_bins=1025):
+
+        super(StretchSpecTime, self).__init__()
+
+        self.rate = rate
+        phi_advance = torch.linspace(
+            0, math.pi * hop_len, num_bins)[..., None]
+
+        self.register_buffer('phi_advance', phi_advance)
```
[the remark this answers is not in the document]
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: it's the same used in `librosa`... maybe `phase_advance` better?

File: torchaudio_contrib/augmentation.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        return phase_vocoder(spect, rate, self.phi_advance)
+
+    def __repr__(self):
+        param_str = '(rate={})'.format(self.rate)
+        return self.__class__.__name__ + param_str
+
+
+class GaussianNoise(_ModuleNoStateBuffers):
+    """
+    Apply gaussian noise to a spectrogram.
```
[the remark this answers is not in the document]
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: > later it actually turns out that it doesn't really apply anything to the original input. if the number of operations (add, multiply,..) are small, i think we can simply have all layers for them for an easier use like

Yeah actually in the latest commit I had fixed that and it did add noise to the spectogram, first commit had a typo my bad. 

>For spectrogram/sound, probably the operation is fairly limited to add, and it's more interesting to have different types of noises (white noise==gaussian, then brown noise, pink noise..).

Yeah that's fair... It's not really an issue with the white noise, but could we do brown and pink noise directly on the spectrogram? If not we could have the `AdditiveNoise()` augmentations on the waveform and then do ` nn.Sequential([AdditiveNoise(), *Spectrogram()])` ...

File: torchaudio_contrib/augmentation.py (hunk 3 above)
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: Yeah definitley, I was using `spect` as a placeholder until we have some consensus on the whole `mag_` / `real_` / `_gram` / `specgrams` debate! Once we do we'll have to refactor `layers` and `functional` too.

Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: Which reminds me per #47 to change the file structure soon.

File: torchaudio_contrib/augmentation.py (hunk 5 above)
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: How about `along_batch`? `same_along_batch` might be too verbose..

File: torchaudio_contrib/augmentation.py (hunk 4 above)
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: >also i think we need little discussion at some point about if this masking would be always with zeros (which i think people would assume) or with a number as implemented here.

So the functional doesn't have a default but in the layer I've set the default to 0 (as users would expect) while also allowing to set a desired value in case the spectrogram isn't whitened.

> masking the masked region is not valid values with zeros, but just doesn't exist at all

Oh right... Yeah I this seems different than `SpecAug` but could be done with something like `torch.masked_select`? Although this would have to be used very carefully when fed into a network... (keeping track of lengths and heights, padding, etc.) but cool.

File: torchaudio_contrib/augmentation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: Yeah, actually I don't really know much about the mechanism. But `phase_advance` sounds better, if it's correct.

File: torchaudio_contrib/augmentation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: Actually, brown and pink noise directly on spectrogram would be easier than generating them on time domain. But, to be honest, I don't think we need to decide it at this moment, not sure if anyone would really need it at any point.

File: torchaudio_contrib/augmentation.py (hunk 3 above)
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: 👍

File: torchaudio_contrib/augmentation.py (hunk 4 above)
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: > (keeping track of lengths and heights, padding, etc.)

Ugh, I missed that. That would be very very ugly, your implementation should be the way to go IMO.

File: torchaudio_contrib/augmentation.py (hunk 5 above)
Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: or, `mask_per_sample`? `per_sample`? hm, not easy :o `batchwise_mask`? ...

Comment (keunwoochoi) [reply torchaudio_contrib/augmentation.py]: On things like this, prob we can simply add a comment saying the name doesn't sound optimal and proceed for the moment.

File: torchaudio_contrib/augmentation.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ksanjeevan) [reply torchaudio_contrib/augmentation.py]: Ok, do you think we should remove it then? I know just summing gaussian noise is kinda trivial but if it's a useful augmentation I think there's no harm in having it as an aug layer..

**Comment by cpuhrsch** [comment]
We could also really need this in torchaudio :)

**Comment by ksanjeevan** [comment]
@cpuhrsch I think once we've had the layers discussion, wave/spect augmentation seems like a useful next step! Plus with the resampling done, we now have implementations for TimeStretch, PitchShift, and Time/Pitch masking :)
